src/code-stats-insights.ts

```diff
@@ -1,121 +1,209 @@
+ *
+ * 1. Old API (only one code stats insight can live for entire setting cascade)
+ *
+ * "codeStatsInsights.query": "repo:^github\\.com/sourcegraph/sourcegraph$",
+ * "codeStatsInsights.otherThreshold": 0.01,
+ *
+ * 2. New API (any number of stats insight can be create in user/org setting cascade)
+ *
+ * "codeStatsInsights.insight.sourcegraphLanguageUsage": {
+ *      "title": "Sourcegraph Language Usage",
+ *      "repository": "github.com/sourcegraph/sourcegraph",
+ *      "otherThreshold": "0.03"
+ * }
+ * */
 export function activate(context: sourcegraph.ExtensionContext): void {
-    const provideView = ({ viewer }: { viewer?: sourcegraph.DirectoryViewer }) => {
-        return from(sourcegraph.configuration).pipe(
-            map(() => sourcegraph.configuration.get().value),
-            switchMap(configuration => {
-                if (!configuration['codeStatsInsights.query']) {
-                    return []
-                }
+    const settings = from(sourcegraph.configuration).pipe(
+        startWith(null),
+        map(() => sourcegraph.configuration.get().value)
+    )
+
+    // Observe stats insights settings from user/org setting cascade
+    const insightChanges = settings.pipe(
+        map(settings => {
+                const insightsFromCreationFlow = Array.from(Object.entries(settings))
+                    .filter(([key]) =>
+                        key.startsWith('codeStatsInsights.insight.')) as [string, Insight | null | false][];
+
+                // In a first version of this extension we had different approach hot to set settings
+                // for the sake of backward compatibility we support this old API as well here
+                const insightFromOldAPI: [string, Insight | null] = [
+                    'codeStatsInsights.insight.language',
+                    settings['codeStatsInsights.query']
+                        ? {
+                            title: 'Language usage',
+                            query: settings['codeStatsInsights.query'],
+                            otherThreshold: settings['codeStatsInsights.otherThreshold']
+                          } as Insight
+                        : null
+                ]
+
+
+                return [insightFromOldAPI, ...insightsFromCreationFlow,]
+            }
+        ),
+        distinctUntilChanged((a, b) => isEqual(a, b))
+    )
+
+    // Further in main context.subscriptions we have nested calls of context.subscriptions
+    // as well, because of this we have to manage inner subscription ourselves. This
+    // subscription bag needed to unsubscribe nested calls in a moment when user/org settings
+    // have been updated.
+    let previousSubscriptions = new Subscription();
+
+    context.subscriptions.add(
+        insightChanges.pipe(
+            tap(() => {
+                previousSubscriptions.unsubscribe()
+                previousSubscriptions = new Subscription();
```

Comment:
**suggestion**: you could add this bag to the extension bag here, so that you don't need to add the individual ones to the outer bag as well.
```suggestion
                previousSubscriptions = new Subscription();
                context.subscriptions.add(previousSubscriptions)
```

Reply:
Aha, I see the commet below, got it